java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -21,336 +21,312 @@
-      symmetric = binaryRelation(kind.symmetric(), rightOp, leftOp);
+  private static BinaryRelation combineTransitively(BinaryRelation a, BinaryRelation b) {
+    if (isEqualityRelation(a.kind) || isEqualityRelation(b.kind)) {
+      return equalityTransitiveBuilder(a, b);
     }
-    return symmetric;
-  }
-
-  /**
-   * @param relation a relation between symbolic values
-   * @return a RelationState<ul>
-   * <li>FULFILLED  if the receiver implies that the supplied relation is true</li>
-   * <li>UNFULFILLED if the receiver implies that the supplied relation is false</li>
-   * <li>UNDETERMINED  otherwise</li>
-   * </ul>
-   * @see RelationState
-   */
-  protected RelationState implies(BinaryRelation relation) {
-    if (leftOp.equals(relation.leftOp) && rightOp.equals(relation.rightOp)) {
-      return relation.isImpliedBy(this);
-    } else if (leftOp.equals(relation.rightOp) && rightOp.equals(relation.leftOp)) {
-      return relation.symmetric().isImpliedBy(this);
+    if (a.kind == LESS_THAN) {
+      return lessThanTransitiveBuilder(a, b);
     }
-    return RelationState.UNDETERMINED;
+    if (b.kind == LESS_THAN) {
+      return lessThanTransitiveBuilder(b, a);
+    }
+    if (a.kind == GREATER_THAN_OR_EQUAL && b.kind == GREATER_THAN_OR_EQUAL) {
+      BinaryRelation greaterThanEqual = greaterThanEqualTransitiveBuilder(a, b);
+      return greaterThanEqual != null ? greaterThanEqual : greaterThanEqualTransitiveBuilder(b, a);
+    }
+    return null;
   }
 
-  protected abstract RelationState isImpliedBy(BinaryRelation relation);
-
-  protected abstract RelationState impliesEqual();
-
-  protected abstract RelationState impliesNotEqual();
-
-  protected abstract RelationState impliesMethodEquals();
-
-  protected abstract RelationState impliesNotMethodEquals();
-
-  protected abstract RelationState impliesGreaterThan();
-
-  protected abstract RelationState impliesGreaterThanOrEqual();
-
-  protected abstract RelationState impliesLessThan();
-
-  protected abstract RelationState impliesLessThanOrEqual();
+  private static boolean isEqualityRelation(Kind kind) {
+    return kind == EQUAL || kind == METHOD_EQUALS;
+  }
 
-  /**
-   * Returns a new relation resulting of the transitive combination of the receiver with the supplied relation.
-   * Schema aRb & bSc => aTc
-   * where aRb is the relation described by the receiver, bSc that of the supplied relation, aTc is the returned relation.
-   * 
-   * @param relation another relation whose left operand is the same as the receiver's right operand
-   * @return the resulting relation or null if the conjunction of both relations does not bring any new information
-   */
-  @CheckForNull
-  protected abstract BinaryRelation combinedAfter(BinaryRelation relation);
+  private static BinaryRelation equalityTransitiveBuilder(BinaryRelation a, BinaryRelation b) {
+    BinaryRelation equality = isEqualityRelation(a.kind) ? a : b;
+    BinaryRelation other = equality == a ? b : a;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)